_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+
+This solution is not ideal, because we have to trust the client of this class to run the `isInactive()` method with the `checkKeepAliveInterval` interval. On the other hand, the `Connection` class won’t be visible to the end user of our `SocketClient` so I think we can live with that.
+
+#### Connections
+Now we need a container from which we can lease connections.
+``` kotlin
+private class Connections {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(Connections::class.java)
+    }
+
+    private val leased = ConcurrentHashMap<ConnectionDestination, BlockingDeque<Connection>>()
+    private val available = ConcurrentHashMap<ConnectionDestination, BlockingDeque<Connection>>()
+
+    fun addLeased(connection: Connection) { // (2)
+        leased(connection.destination) += connection
+    }
+
+    fun lease(destination: ConnectionDestination): Connection? { // (1)
+        // There can be already closed connection, but we want to return an active one
+        do {
+            val connection = available(destination).poll()
+            if (connection == null) {
+                return null
+            } else if (!connection.isClosed()) {
+                addLeased(connection)
+                return connection
+            }
+        } while(true)
+    }
+
+    fun release(connection: Connection) { // (3)
+        leased(connection.destination).remove(connection)
+        available(connection.destination) += connection
+    }
+
+    fun closeStale() { // (4)
+        closeStale(leased)
+        closeStale(available)
+    }
+
+    private fun leased(destination: ConnectionDestination) =
+            leased.computeIfAbsent(destination) { LinkedBlockingDeque() }
+
+    private fun available(destination: ConnectionDestination) =
+            available.computeIfAbsent(destination) { LinkedBlockingDeque() }
+
+    private fun closeStale(
+            connectionsForDestination: ConcurrentHashMap<ConnectionDestination, BlockingDeque<Connection>>) {
+        connectionsForDestination.forEach { _, connections ->
+            val staleConnections = connections.filter(Connection::isInactive)
+            connections.removeAll(staleConnections)
+            staleConnections.forEach {
+                logger.debug("Closing connection to ${it.destination} due to not being active")
+                it.close()
+            }
+        }
+    }
+
+}
+```
+We keep separate queues of connections for leased and available connections. We can lease a connection (1), but we have to be careful not to return a closed connection. We can add already leased connection (2). This will be useful when we will create a connection and return it immediately to the client. We can release connection (3) so it is placed back into the leased queue.
+There is also a method for closing stale connections (4), it seeks for inactive connections, closes it and removes those from the queues.
```

Comment:
There is also **a method** for closing stale connections (4), **it** seeks for inactive connections, closes **it** and removes **those** from the queues.

First *it* refers to "a method", but I don't know what the second *it* refers to. Presumably "inactive connections", but they are plural, so it should be "them". "Those" I don't think should be in there.

I'd write:
There is also a method for closing stale connections (4), it seeks for inactive connections, closes them and removes from the queues.

(you've already introduced the queues couple of sentences above so it should be fine)

Reply:
You are right, this sounds confusing.